### PR TITLE
TopSearchBar: Refactoring, simplifications and optimizations

### PR DESCRIPTION
**Pull request by torkelo**
Trying to reduce the amount of re-work on re-render and the causes of re-renders in the TopSearchBar. Since this component is on every single page optimizing re-renders has an impact on all pages. Also simplifies code.

Cause of re-renders
* Any location / search change
* Any search query change (could cause double re-renders as it also changes location)

Changes
* By moving search into a separate component we can skip re-rendering whole TopSearchBar on query change (and simplifies this component)
* Leverage that Dropdown supports lazy rendering of the overlay. Move all the logic of enriching and adding tracking to nav items to when dropdown is opened
* Move the opening of the OrgSwitching to the enriching function and handle it the same way as the KeyboardShortcuts modal

### CODE REVIEW

File: public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx [hunk 1]
```diff
@@ -1,18 +1,25 @@
...
 export interface TopNavBarMenuProps {
   node: NavModelItem;
 }
 
-export function TopNavBarMenu({ node }: TopNavBarMenuProps) {
+export function TopNavBarMenu({ node: nodePlain }: TopNavBarMenuProps) {
   const styles = useStyles2(getStyles);
+  const location = useLocation();
+  const enriched = enrichConfigItems([cloneDeep(nodePlain)], location);
+  const node = enrichWithInteractionTracking(enriched[0], false);
```
Comment (torkelo) [inline public/app/core/components/AppChrome/TopBar/TopNavBarMenu.tsx:21]: now useLocation, enrich + tracking is only done for a single tree node. And this component is only rendered when the Dropdown overlay is opened

File: public/app/core/components/AppChrome/TopSearchBar.tsx [hunk 2]
```diff
@@ -56,17 +27,11 @@ export function TopSearchBar() {
...
-          value={query.query ?? ''}
-          onChange={onSearchChange}
-          className={styles.searchInput}
-        />
+        <TopSearchBarInput />
       </div>
       <div className={styles.actions}>
         {helpNode && (
-          <Dropdown overlay={<TopNavBarMenu node={helpNode} />}>
+          <Dropdown overlay={() => <TopNavBarMenu node={helpNode} />}>
```
Comment (lpskdl) [inline public/app/core/components/AppChrome/TopSearchBar.tsx:34]: Hi @torkelo, I would like to know the idea why this is changed to anonymous function? thanks

**Review by lpskdl — APPROVED** [review]
Thanks for refactoring. It's really clean now. Tested on my side and it works.

File: public/app/core/components/AppChrome/TopSearchBar.tsx (hunk 2 above)
Comment (torkelo) [reply public/app/core/components/AppChrome/TopSearchBar.tsx:34]: @lpskdl this is what makes the menu "lazy" rendered only when it is opened.

**Review by ashharrison90 — APPROVED** [review]
yep, this is much cleaner 😍

**Pull request merged**